src/hal/hal_i2c_microSpecific.c

```diff
@@ -0,0 +1,77 @@
+#include "hal.h"
+#include "hal_i2c.h"
+
+#include "hal_i2c_microSpecific.h"
+#include "stm32l031xx.h"
+#include "stm32l0xx.h"
+#include "system_stm32l0xx.h"
+
+#include "shared_defs.h"
+
+
+hal_error_E hal_i2c_microSpecific_init(void) {
+
+    //This is to configure it in master mode
+
+    //clear PE bit
+    I2C1->CR1 &= ~(1<<I2C_CR1_PE_Pos);
+
+    //Enable analog filter
+    I2C1->CR1 &= ~(1<<I2C_CR1_ANFOFF_Pos);
+
+    //Disable digital filter
+    I2C1->CR1 &= ~(I2C_CR1_DNF_Msk);
+
+    //Configure timings
+    //Generated with CubeMX might not be best practice
+    I2C1->TIMINGR = 0x00B07CB4;
+
+    //NOSTRETCH must be kept clear in master mode
+    I2C1->CR1 &= ~(1<<I2C_CR1_NOSTRETCH_Pos);
+
+    //Enable the peripheral
+    I2C1->CR1 |= 1 << I2C_CR1_PE_Pos;
+}
+
+hal_error_E hal_i2c_sendData(uint8_t* data, uint8_t nBytes) {
+
+    //Set number of bytes to be received
+    I2C1->CR2 &= nBytes<<I2C_CR2_NBYTES_Pos;
+
+    for (uint8_t i = 0; i < nBytes; i++) {
+
+        //Wait until TXDR is empty
+        while (!(I2C1->ISR & I2C_ISR_TXE_Msk));
+
+        //Load next byte into TXDR
+        I2C1->TXDR = *data;
+        data++;
+    }
+}
+
+hal_error_E hal_i2c_receiveData(uint8_t* data, uint8_t nBytes) {
+
+    //Set number of bytes to be received
+    I2C1->CR2 &= nBytes<<I2C_CR2_NBYTES_Pos;
+
+    for (uint8_t i = 0; i < nBytes; i++) {
+        while (!(I2C1->ISR & I2C_ISR_RXNE_Msk));
```

Comment:
Did a quick read on this and it seems that there is no hardware buffering. So for your initial testing, run this outside of FreeRTOS since the added scheduler latency may lead to dropped bytes. You can also check me by trying it in a freeRTOS task of varying tick rate as well. My guess is that we will need an interrupt to grab bytes and place them in a buffer for us. (but we can cross that bridge later)